pkg/controller/machine/machine.go

```diff
@@ -555,6 +547,7 @@ func (c *Controller) deleteCloudProviderInstance(prov cloud.Provider, machine *c
 	}
 
 	machine, err = c.updateMachine(machine, func(m *clusterv1alpha1.Machine) {
+		finalizers := sets.NewString(m.Finalizers...)
```

Comment:
Re-using the `finalizers` set we created before calling the providers `Cleanup` is pretty problematic, because:

* We pass `machineCreateDeleteData` into `Cleanup` which contains a `MachineUpdater`
* This allows providers to Delete provider-specific finalizers
* If that happened, we will re-add them because we use a `finalizers` set that was created _before_ we called `Cleanup`

Since this may result in a no-op updateFunc, I have changed `updateMachine` to be able to cope with that case